File: 0238-product-of-array-except-self/0238-product-of-array-except-self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
//     let leftArr = [];
//     let rightArr = [];
//     let leftMulti = 1;
//     let rightMulti = 1;
    
//     for(let i=0; i<nums.length; i++){
//         leftArr[i] = leftMulti;
//         leftMulti *= nums[i];
//     }
    
//     for(let i=nums.length-1; i>=0; i--){
//         rightArr[i] = rightMulti;
//         rightMulti *= nums[i];
//         rightArr[i] *= leftArr[i]
//     }
//     return rightArr;
    
    let ans = [];
    let multi = 1;
    
    for(let i=0; i<nums.length; i++){
        ans[i] = multi;
        multi *= nums[i];
    }
    
    multi = 1;
    for(let i=nums.length-1; i>=0; i--){
        ans[i] *= multi;
        multi *= nums[i];
        
    }
    return ans;
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
        length = len(nums)
        left, right, ans = [0]*length,[0]*length,[0]*length 
        left[0] = 1
        
        for i in range(1, length ):
            left[i] = left[i-1] * nums[i-1]
            
        right[length-1] = 1
        
        for i in reversed(range(length-1)):
            right[i] = right[i+1] * nums[i+1]
            
        for i in range(length):
            ans[i] = left[i] * right[i]
            
        return ans
        
#         pre = 1
#         ans = [1] * len(nums)
#         for i in range(len(nums)):
#             ans[i] = pre
#             pre *= nums[i]
            
        
#         post = 1
#         for i in reversed(range(len(nums))):
#             ans[i] *= post
#             post *= nums[i]
            
#         return ans
        
#         length = len(nums)
#         left, right, ans = [0]*length, [0]*length, [0]*length
        
#         left[0] = 1
        
#         for i in range(1, length):
#             left[i] = left[i-1] * nums[i-1]
            
#         right[length-1] = 1
        
#         for i in reversed(range(length-1)):
#             right[i] = right[i+1]*nums[i+1]
            
#         for i in range(len(nums)):
#             ans[i] = left[i] * right[i]
            
#         return ans

        